File: app/src/main/java/com/tunjid/androidbootstrap/viewmodels/DoggoRankViewModel.java
package com.tunjid.androidbootstrap.viewmodels;

import android.app.Application;
import android.content.Context;
import android.util.Pair;

import com.tunjid.androidbootstrap.R;
import com.tunjid.androidbootstrap.functions.TransformingSequentialList;
import com.tunjid.androidbootstrap.model.Doggo;
import com.tunjid.androidbootstrap.recyclerview.Differentiable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.recyclerview.widget.DiffUtil;
import io.reactivex.Flowable;
import io.reactivex.Single;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.processors.PublishProcessor;

import static androidx.recyclerview.widget.ItemTouchHelper.ACTION_STATE_IDLE;
import static androidx.recyclerview.widget.ItemTouchHelper.ACTION_STATE_SWIPE;
import static io.reactivex.android.schedulers.AndroidSchedulers.mainThread;
import static io.reactivex.schedulers.Schedulers.io;

public class DoggoRankViewModel extends AndroidViewModel {

    private Pair<Long, Integer> doggoIdPositionPair;
    private final List<Doggo> doggos;
    private final CompositeDisposable disposables;
    private final PublishProcessor<DiffUtil.DiffResult> processor;

    public DoggoRankViewModel(@NonNull Application application) {
        super(application);
        doggos = new ArrayList<>(Doggo.doggos);
        disposables = new CompositeDisposable();
        processor = PublishProcessor.create();
    }

    public List<Doggo> getDoggos() {
        return doggos;
    }

    public Flowable<DiffUtil.DiffResult> watchDoggos() {
        return processor;
    }

    public void onActionStarted(Pair<Long, Integer> doggoIdActionPair) {
        long id = doggoIdActionPair.first;
        int currentIndex = TransformingSequentialList.transform(doggos, Doggo::hashCode).indexOf((int) id);
        doggoIdPositionPair = new Pair<>(id, currentIndex);
    }

    public void swap(int from, int to) {
        if (from < to) for (int i = from; i < to; i++) Collections.swap(doggos, i, i + 1);
        else for (int i = from; i > to; i--) Collections.swap(doggos, i, i - 1);
    }

    public Pair<Integer, Integer> remove(int position) {
        doggos.remove(position);

        int lastIndex = doggos.size() - 1;
        return new Pair<>(Math.min(position, lastIndex), lastIndex);
    }

    public String onActionEnded(Pair<Long, Integer> doggoIdActionPair) {
        int action = doggoIdActionPair.second;
        int startPosition = doggoIdPositionPair.second;
        long startId = doggoIdPositionPair.first;
        long endId = doggoIdActionPair.first;

        if (action == ACTION_STATE_IDLE || startId != endId) return "";

        boolean isRemoving = action == ACTION_STATE_SWIPE;
        List<Integer> ids = TransformingSequentialList.transform(isRemoving ? Doggo.doggos : doggos, Doggo::hashCode);
        int endPosition = ids.indexOf((int) endId);

        if (endPosition < 0) return "";
        Doggo doggo = (isRemoving ? Doggo.doggos : doggos).get(endPosition);

        if (isRemoving && doggos.indexOf(doggo) >= 0) return ""; // Doggo is still in the list
        else if (!isRemoving && startPosition == endPosition) return ""; // Doggo kept it's rank

        Context context = getApplication();
        if (isRemoving) return context.getString(R.string.doggo_removed, doggo.getName());
        else return context.getString(R.string.doggo_moved, doggo.getName(), endPosition + 1);
    }

    public void resetList() {
        disposables.add(Single.fromCallable(() -> Differentiable.diff(
                doggos,
                Doggo.doggos,
                (destination, additions) -> additions,
                doggo -> Differentiable.fromCharSequence(() -> String.valueOf(doggo.hashCode()))))
                .subscribeOn(io())
                .observeOn(mainThread())
                .subscribe(doggoDiff -> {
                    doggos.clear();
                    doggos.addAll(doggoDiff.cumulative);
                    processor.onNext(doggoDiff.result);
                }));
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        disposables.clear();
        processor.onComplete();
    }
}

File: app/src/main/java/com/tunjid/androidbootstrap/viewmodels/DoggoRankViewModel.kt
package com.tunjid.androidbootstrap.viewmodels

import android.app.Application
import android.util.Pair
import androidx.lifecycle.AndroidViewModel
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ItemTouchHelper.ACTION_STATE_IDLE
import androidx.recyclerview.widget.ItemTouchHelper.ACTION_STATE_SWIPE
import com.tunjid.androidbootstrap.R
import com.tunjid.androidbootstrap.functions.collections.Lists
import com.tunjid.androidbootstrap.model.Doggo
import com.tunjid.androidbootstrap.recyclerview.diff.Diff
import com.tunjid.androidbootstrap.recyclerview.diff.Differentiable
import io.reactivex.Flowable
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers.mainThread
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.processors.PublishProcessor
import io.reactivex.schedulers.Schedulers.io
import java.util.*
import kotlin.math.min

class DoggoRankViewModel(application: Application) : AndroidViewModel(application) {

    val doggos: MutableList<Doggo> = ArrayList(Doggo.doggos)
    private val disposables: CompositeDisposable = CompositeDisposable()
    private val processor: PublishProcessor<DiffUtil.DiffResult> = PublishProcessor.create()

    private var doggoIdPositionPair: Pair<Long, Int>? = null

    fun watchDoggos(): Flowable<DiffUtil.DiffResult> = processor

    fun onActionStarted(doggoIdActionPair: Pair<Long, Int>) {
        val id = doggoIdActionPair.first
        val currentIndex = Lists.transform(doggos) { it.hashCode() }.indexOf(id.toInt())
        doggoIdPositionPair = Pair(id, currentIndex)
    }

    fun swap(from: Int, to: Int) {
        if (from < to) for (i in from until to) Collections.swap(doggos, i, i + 1)
        else for (i in from downTo to + 1) Collections.swap(doggos, i, i - 1)
    }

    fun remove(position: Int): Pair<Int, Int> {
        doggos.removeAt(position)

        val lastIndex = doggos.size - 1
        return Pair(min(position, lastIndex), lastIndex)
    }

    fun onActionEnded(doggoIdActionPair: Pair<Long, Int>): String {
        val action = doggoIdActionPair.second
        val startPosition = doggoIdPositionPair!!.second
        val startId = doggoIdPositionPair!!.first
        val endId = doggoIdActionPair.first

        if (action == ACTION_STATE_IDLE || startId != endId) return ""

        val isRemoving = action == ACTION_STATE_SWIPE
        val ids = Lists.transform(if (isRemoving) Doggo.doggos else doggos) { it.hashCode() }
        val endPosition = ids.indexOf(endId.toInt())

        if (endPosition < 0) return ""
        val doggo = (if (isRemoving) Doggo.doggos else doggos)[endPosition]

        if (isRemoving && doggos.indexOf(doggo) >= 0) return "" // Doggo is still in the list
        else if (!isRemoving && startPosition == endPosition) return "" // Doggo kept it's rank

        val context = getApplication<Application>()

        return if (isRemoving) context.getString(R.string.doggo_removed, doggo.name)
        else context.getString(R.string.doggo_moved, doggo.name, endPosition + 1)
    }

    fun resetList() {
        disposables.add(Single.fromCallable {
            Diff.calculate(
                    doggos,
                    Doggo.doggos,
                    { _, additions -> additions },
                    { doggo -> Differentiable.fromCharSequence { doggo.hashCode().toString() } })
        }
                .subscribeOn(io())
                .observeOn(mainThread())
                .subscribe { doggoDiff ->
                    doggos.clear()
                    doggos.addAll(doggoDiff.items)
                    processor.onNext(doggoDiff.result)
                })
    }

    override fun onCleared() {
        super.onCleared()
        disposables.clear()
        processor.onComplete()
    }
}
